frontend-cert/js-projects/music-player-lab/script.js

```diff
@@ -0,0 +1,217 @@
+  songs: allSongs,
+  currentSong: null,
+  songCurrentTime: 0,
+};
+
+const playSong = (id, start = true) => {
+  const song = userData.songs.find((song) => song.id === id);
+  audio.src = song.src;
+  audio.title = song.title;
+
+  if (
+    userData.currentSong === null ||
+    playButton.classList.contains("playing") ||
+    start
+  ) {
+    audio.currentTime = 0;
+  } else {
+    audio.currentTime = userData.songCurrentTime;
+  }
+  userData.currentSong = song;
+  playButton.classList.add("playing");
+
+  highlightCurrentSong();
+  setPlayerDisplay();
+  setPlayButtonAccessibleText();
+  audio.play();
+};
+
+const pauseSong = () => {
+  userData.songCurrentTime = audio.currentTime;
+  playButton.classList.remove("playing");
+  audio.pause();
+};
+
+const getCurrentSongIndex = () => userData.songs.indexOf(userData.currentSong);
+
+const getNextSong = () => userData.songs[getCurrentSongIndex() + 1];
+
+const getPreviousSong = () => userData.songs[getCurrentSongIndex() - 1];
+
+const playNextSong = () => {
+  if (userData.currentSong === null) {
+    playSong(userData.songs[0].id);
+    return
+  }
+  const nextSong = getNextSong();
+  if (nextSong) {
+    playSong(nextSong.id);
+  } else {
+    userData.currentSong = null;
+    userData.songCurrentTime = 0;
+    pauseSong();
+    setPlayerDisplay();
+    highlightCurrentSong();
+    setPlayButtonAccessibleText();
+  }
+};
+
+const playPreviousSong = () => {
+  if (userData.currentSong === null) return;
+  const previousSong = getPreviousSong();
+  if (previousSong) {
+    playSong(previousSong.id);
+  } else {
+    playSong(userData.songs[0].id);
+  }
+};
+
+const setPlayerDisplay = () => {
+  const playingSong = document.getElementById("player-song-title");
+  const songArtist = document.getElementById("player-song-artist");
+  const currentTitle = userData.currentSong?.title;
+  const currentArtist = userData.currentSong?.artist;
+
+  playingSong.textContent = currentTitle ? currentTitle : "";
+  songArtist.textContent = currentArtist ? currentArtist : "";
+};
+
+const highlightCurrentSong = () => {
+  const playlistSongElements = document.querySelectorAll(".playlist-song");
+  const songToHighlight = document.getElementById(
+    `song-${userData.currentSong?.id}`
+  );
+
+  playlistSongElements.forEach((songEl) => {
+    songEl.removeAttribute("aria-current");
+  });
+
+  if (songToHighlight) songToHighlight.setAttribute("aria-current", "true");
+};
+
+const renderSongs = (array) => {
+  const songsHTML = array
+    .map((song) => {
+      return `
+      <li id="song-${song.id}" class="playlist-song">
+      <button class="playlist-song-info" onclick="playSong(${song.id})">
+          <span class="playlist-song-title">${song.title}</span>
+          <span class="playlist-song-artist">${song.artist}</span>
+          <span class="playlist-song-duration">${song.duration}</span>
+      </button>
+      </li>
+      `;
+    })
+    .join("");
+
+  playlistSongs.innerHTML = songsHTML;
+};
```

Comment:
I haven't thought about that. Not ideal to have the playlist hardcoded in the HTML but I guess we can do that if we want to shorten the workshop.

Reply:
I am not sure why it would be bad to have the playlist hardcoded.
Is there an advantage to rendering the songs in javascript? because it seems to work well in the HTML